provider/file/file.go

```diff
@@ -18,68 +22,149 @@ var _ provider.Provider = (*Provider)(nil)
 // Provider holds configurations of the provider.
 type Provider struct {
 	provider.BaseProvider `mapstructure:",squash"`
+	Directory             string `description:"Load configuration from one or more .toml files in a directory"`
 }
 
 // Provide allows the file provider to provide configurations to traefik
 // using the given configuration channel.
 func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
-	watcher, err := fsnotify.NewWatcher()
+	var configuration *types.Configuration
+	var err error
+	var watchDir string
+
+	if p.Directory != "" {
+		watchDir = p.Directory
+		configuration, err = loadFileConfigFromDirectory(p.Directory)
+	} else {
+		watchDir = filepath.Dir(p.Filename)
+		configuration, err = loadFileConfig(p.Filename)
+	}
```

Comment:
I think you can replace this duplicate block (here and in the ` watcherCallback`) by this method:

```golang
func (p *Provider) loadConfig() (*types.Configuration, error) {
	if p.Directory != "" {
		return loadFileConfigFromDirectory(p.Directory)
	} else {
		return loadFileConfig(p.Filename)
	}
}
```